### DISCUSSION TITLE
Update images parallely to Git

### DISCUSSION THREAD

**Discussion by not-ajay** [fluxcd/flux2#2973] [Q&A] [answered]
We have multiple services which use their own image repositories of them, we push new images to all image repositories in short amount of time.
Flux scans the image repositories and recongise as latest image but flux updates images one by one with single commit each time which takes much time:
Can flux update images parallely ?
Can flux update multiple images in single commit with same time as time taken by single commit with 1 image?

**Comment by kingdonb** [comment, marked as answer]
Flux reconciles ImageRepository one by one and as each is triggered, the policy which subscribes to it will be notified. When the ImageUpdateAutomation reconciles on its `spec.interval`, any ImageRepository which has received an update will be ready for IUA to trigger a git commit from the policy, and it follows that any ImageRepository which did not reconcile already and trigger to update the policy that belongs to it, would have been left behind by ImageUpdateAutomation's next commit.

So without trying to sway you away from this perfectly good feature part of the Image Automation API too much, there are different alternatives you might want to consider still in Flux; some of them involving the ImageUpdateAutomation, and others which can avoid it entirely that are brand new. Here's a CNCF webinar we gave last week where this new OCI automation solution approach is first presented:

[Cloud Native Live: VSCode and Flux: Testing the new OCI Repository feature](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cloud-native-live-vscode-and-flux-testing-the-new-unreleased-oci-repository-feature/) (~un~released! No longer unreleased...)


A small change without upheaving everything that might still work well for you is if ImageUpdateAutomation was set with a longer `spec.interval` – definitely at least 5m, maybe longer like 15m, so that updates take a little longer to arrive, but it's unlikely you'd ever see the Automation commit running on an interval that's physically between the time of publish of images that were meant to go out together.

Then, make sure every ImageRepository is reconciled frequently enough that all the images that updated will be ready before that 15m is up. This is not a perfect solution and I am not sure if ImageUpdateAutomation might have somehow subscribed through the Policy. That would foil this approach from working. (There is a chance that ImageUpdateAutomations automatically subscribe to any ImagePolicy in their own namespace, I'm unfortunately not sure.)

You could try to orchestrate all this if your only concern now is too much noise on the repo, but really if you need to update a group of images in parallel with atomicity, this is beginning to sound like "atomic releases" which are a problem we first solved with Helm, and later Kustomize controller (`spec.wait` which can revert and rollback a release with a part that fails). And as I mentioned at the start, now Flux has a brand new solution for that approach which I think you might want to take a look at (the `OCIRepository` source kind.)

Just released with Flux 0.32 is the [OCIRepository manifest source](https://fluxcd.io/docs/cheatsheets/oci-artifacts/), which is a natural fit for this use case. Instead of using ImageUpdateAutomation to push out the manifest changes, use CI to do it, but pushing the updated manifests not back to Git, but ahead forward to an OCIRepository for storing versions of the application's manifests. Another alternative is doing the update manually with a tool like Makefile then commit this back to Git before the next release. Either one is good. Then, we think for most users they will choose to either use a `production` tag on the OCI repo like the `main` branch of your Flux Git repo, or the familiar semver notation in the OCIRepository sources that select the latest manifest by semver tag version.

This is the approach which was demonstrated on the YouTube link from CNCF that I posted up above. ☝️ 

Or bring in an ImageUpdateAutomation, and use it there if you prefer to select the last released version and make it more convincingly auditable with that Git commit. I know there are probably a lot of people out there who want to have exactly which version was deployed in production at any given time in a Git commit which is unambiguous and some cryptographic proof of integrity is possible.

Now Flux can manage those types of deployments and make atomic releases even without Helm, in an automated way. It only requires a bit of help from CI and the flux binary in your CI pipeline, which you can get via GitHub Actions easily 👍
